### PR TITLE
MAINT: Simplify npymath

### PR DESCRIPTION
**Pull request by mattip**
Towards solving #20880. So far this removes support for outdated Solaris, rearranges the headers a bit, adds some double C99 routines to mandatory math functions, removes emitting HAVE_* macros for mandatory functions, and removes code paths in the cases of missing mandatory functions.

It seems we cannot remove the complex math function implementations until we drop manylinux2014 (glibc 2.17) which does not [go EOL](https://github.com/mayeut/pep600_compliance#acceptable-distros-to-build-wheels) until June 2024.

### CODE REVIEW

File: numpy/core/setup_common.py
Comment (mattip) [inline numpy/core/setup_common.py:131]: `hypot` and `atan2` are problematic on windows, see npy_config.h for the logic to disable the built-in functions and use the alternatives I removed. I guess I should move them to `OPTIONAL` and restore the code in `npy_math_internal.h.src`.

**Comment by mattip** [comment]
The result of #9567 from 2017 was to blocklist hypot on 32-bit windows. When I rerun the test file after compiling with cl.exe for x86 from VS2019, I get that the floating point flags are not changed:
```
>test_hypot.exe
FPU mode: 589855
If a right triangle has sides inf and 4.0, its hypotenuse is inf
FPU mode: 589855
```

Maybe in the mean time Microsoft fixed the problem? We do have a `check_fpu_mode` fixture on tests that should report when the mode changes

**Comment by rgommers** [comment]
Thanks Matti, this looks like a pretty good start already. Some of the outdated long double formats also stood out to me, but probably only worth touching if they're a pain for moving build systems; otherwise best to wait and deprecate it all in one go later.

**Comment by mattip** [comment]
If we wish to vendor the npmath sources into the wheel, this will require thinking about how to package the build-time generated [`config.h`](https://github.com/rgommers/scipy/blob/vendor-npymath/scipy/_lib/npymath/config.h) (from @rgommers' experiment on vendoring npymath for scipy) file. Maybe we should rename it and put it in the regulare numpy include directory? This is similar to Python's `_sysconfigdata__x86_64-linux-gnu.py` (on linux) file that captures all the `#define` macros when building python

**Comment by rgommers** [comment]
> If we wish to vendor the npmath sources into the wheel, this will require thinking about how to package the build-time generated [`config.h`](https://github.com/rgommers/scipy/blob/vendor-npymath/scipy/_lib/npymath/config.h)

That can't work I think, the whole point of this header is to account for platform/OS/compiler-specific differences, so we can't just ship one `config.h` that works with the rest of the sources.

File: numpy/core/setup_common.py
Comment (rgommers) [inline numpy/core/setup_common.py:127]: As a further cleanup, consider merging the elements of this list into the list right above, and remove `MANDATORY_FUNCS +=`

File: numpy/core/setup.py
Comment (rgommers) [inline numpy/core/setup.py:190]: I had a quick peek at this; it's still the case that CPython does this. There's other defines too, like `HAVE_ASINH`, but those are highly unlikely to contain a different value at numpy build time vs. Python build time. Not sure if this check is still needed, but fine to leave as is.

Comment (rgommers) [reply numpy/core/setup.py:190]: Also: `:wags-finger-at-python:`

File: numpy/core/setup_common.py
Comment (rgommers) [reply numpy/core/setup_common.py:131]: `hypot` yes indeed. `atan2` is only undef'd for `MSC_VER < 1900`, so that is no longer relevant. You could clean it up in `npy_config.h` instead.

**Review by rgommers — COMMENTED** [review]
CI isn't happy yet, but the changes LGTM so far.

**Comment by mattip** [comment]
cygwin still needs a blocklisted `log2`. There is something off with one of the `pow` variants and macos. My first goal is to get CI passing and then clean up the resulting code.

**Comment by mattip** [comment]
also see #22108 to fix a warning about deprecated macOS images.

**Comment by mattip** [comment]
rabbit hole: `powl` is blocklisted for macoOS (and is not working correctly on macOS in this PR). In #8608 from 2017 it was noted that `np.float128(2) ** 16383` incorrectly returns `inf`. PR #15074 was started to fix this but never finished. I wonder if removing the blocklist will solve both the inaccuracy from #8608 and the failure here.

File: numpy/core/setup_common.py
Comment (mattip) [reply numpy/core/setup_common.py:129]: > `atan2` is only undef'd for MSC_VER < 1900

It is also undef'd for `__MINGW32_VERSION`, which we do not test for. `MSC_VER 1900` corresponds to visual studio 2015. Officially its [exended end date](https://docs.microsoft.com/en-us/lifecycle/products/visual-studio-2015) is 2025. Do we have a clear policy for dropping older compilers?

File: numpy/core/setup_common.py
Comment (mattip) [reply numpy/core/setup_common.py]: done

**Comment by mattip** [comment]
I think I have gone about as far as I can go until we can drop manylinux2014 (in 2024), msvc 2015 (in 2025?) and 32-bit mingw (which may have fixed hypot and atan2, but we don't test it). Maybe by then cygwin will have fixed its complex functions as well.

File: numpy/core/setup_common.py
Comment (h-vetinari) [reply numpy/core/setup_common.py:146]: On the scipy-side, the minimum requiremen has moved to vs2019 already, because even vs2019 is already [EOL](https://docs.microsoft.com/en-us/lifecycle/products/visual-studio-2017). The Extended date is paid-only support AFAIK, and since vs2019 & vs2022 are ABI-compatible back to vs2015, I don't personally see a reason to keep supporting such old compilers.

**Comment by h-vetinari** [comment]
> until we can drop [...] msvc 2015 (in 2025?)

Let's please do this much earlier. I doubt that vs2015 support is actually still in working order, and the last blocker for moving away from vs2017 ([AFAIU](https://github.com/numpy/numpy/issues/21346#issuecomment-1101432000)) is now finally about to be solved as well: conda-forge [will move](https://github.com/conda-forge/conda-forge.github.io/blob/main/src/user/announcements.rst#2022-08-11-moving-to-visual-studio-toolchain-vc142) to vs2019+vc142 in a few days.

**Comment by matthew-brett** [comment]
I agree with @h-vetinari - let's drop VS < 2019 .  Does anyone have a good argument not to?

**Comment by rgommers** [comment]
The overflow warning in the complex `abs` test looks like a problem.

**Comment by rgommers** [comment]
Note that any pruning of symbols used in `scipy.special` is going to break already released SciPy versions on macOS arm64: https://github.com/MacPython/scipy-wheels/issues/173#issuecomment-1214902006. I don't think that's a hard blocker, as long as we've done one release of SciPy with a fix before the next NumPy release.

File: numpy/core/setup_common.py
Comment (rgommers) [inline numpy/core/setup_common.py:131]: duplicate entries here

**Comment by mattip** [comment]
> Note that any pruning of symbols used in scipy.special is going to break already released SciPy versions

This PR should not remove any of the exported functions in `npymath.a`. It does reduce the mandatory functions to the form `double npy_sin(double x) { return sin(x); }, and removes the `#define HAVE_SIN` define. but as far as I can tell scipy does not use any of the `HAVE_` macros in `scipy.special`.

**Comment by rgommers** [comment]
I have started on the Meson build of `npymath`, so far I've got the checks to generate most of `config.h` implemented (with mandatory funcs based on this PR): https://github.com/numpy/numpy/compare/main...rgommers:numpy:meson?expand=1. It's getting a lot more understandable:)

**Review by mattip — COMMENTED** [review]
I think this is ready for review. `npy_math_internal.h.src` is the most complicated change. Most of what it does is to define functions like
```
float npy_sinf(float x) {
    return sinf(x);
}
```

A second pass through npymath could define macros for many of these mandatory functions rather than generating functions for them
```
#define npy_sin sin
#define npy_sinf sinf
#define npy_sinl sinl
```

Some "optional" functions have more correct implementations. These are the same, with a small refactoring tweak.

File: numpy/core/setup_common.py
Comment (rgommers) [reply numpy/core/setup_common.py:145]: So this logic is still a little more complicated than it should be. These still C99 functions, so wouldn't it be easier to simply say here that they are mandatory? The blocklist is not connected directly to what is happening here; there should not be a case of `pow` et al. not existing at all right?

Comment (mattip) [reply numpy/core/setup_common.py:145]: The name `OPTIONAL_STDFUNCS` is legacy. What about `BLOCKLISTABLE` instead?

Comment (rgommers) [reply numpy/core/setup_common.py:145]: Better, but still the question remains why it should be a separate list. I'd say merge it all into MANDATORY, and just add a comment xref'ing the blocklist that is defined elsewhere.

Comment (rgommers) [reply numpy/core/setup_common.py:145]: Maybe I am not expressing myself clearly enough: these functions are part of C99 and simply must exist just like all other mandatory functions. There is no way that `HAVE_POW` & co are not going to be `1`. There is no reason to treat them differently from the other C99 functions, unless I am missing something. The use of `HAVE_POW` in the blocklist can simply be removed. This cannot fail any more than `HAVE_SIN` et al.

Comment (mattip) [reply numpy/core/setup_common.py:145]: > There is no way that HAVE_POW & co are not going to be 1

The macro `HAVE_POW` must exist for the current blocklist mechanism to be able to undefine it. Since `sin` is not blocklist-able, `HAVE_SIN` is not necessary and is not generated in this PR. Maybe it would be more clear if the blocklist worked differently: it would define a `BLOCK_POW`. Then the code would do
```
double pow(double x, double y) {
#ifndef(BLOCK_POW)
    return pow(x, y);
#else 
    // blocklisted implementation
    ...
#endif
};
```

instead of the current
```
double pow(double x, double y) {
#ifdef(HAVE_POW)
    return pow(x, y);
#else 
    // blocklisted implementation
    ...
#endif
};
```

Comment (rgommers) [reply numpy/core/setup_common.py:145]: Yes, that sounds nicer - shorter and clearer. It also helps keep the meson configure output concise (there'll be one line per check, so unnecesary checks are good to cut out).

Comment (mattip) [reply numpy/core/setup_common.py]: Done for the non-complex blocked functions (assuming CI is happy). The complex ones are more ... complex to convert.

Comment (rgommers) [reply numpy/core/setup_common.py]: Thanks! Yeah, I think we can leave complex alone in this PR, because it's a large list.

Comment (rgommers) [reply numpy/core/setup_common.py]: Eventually we'll rip this code out anyway:)

**Comment by mattip** [comment]
#22139 was merged, once #22159 is resolved this is ready for another review

**Comment by rgommers** [comment]
Needs a merge conflict resolved after gh-22159 was merged.

File: numpy/core/setup_common.py
Comment (h-vetinari) [inline numpy/core/setup_common.py:146]: This comment seems our of date now as `npy_config.h` already uses the BLOCK_ variants.

Comment (h-vetinari) [reply numpy/core/setup_common.py:146]: The "make these mandatory" part is also not yet directly possible, as MSVC does not support them yet at all. They do support C99 complex through their own types though, which should make it [possible](https://github.com/numpy/numpy/issues/20880#issuecomment-1210456286) to achieve effectively the same outcome (well, aside from having to carry shim types in numpy that switch accordingly)

File: numpy/core/src/common/npy_config.h
Comment (h-vetinari) [inline numpy/core/src/common/npy_config.h:15]: If we're renaming these variables already, should we take the opportunity to better namespace them? I.e. `NPY_BLOCK_EXPF`. AFAICT that's better practice than having such numpy-specific things in the global macro namespace.

File: numpy/core/src/common/numpyos.c
Comment (h-vetinari) [inline numpy/core/src/common/numpyos.c:777]: Should these now more or less pointless wrapper functions be deprecated somehow...?

Comment (mattip) [reply numpy/core/src/common/numpyos.c:777]: Agreed there is more cleanup that can be done to turn one-line functions into defines (if they are exported) or to deprecate them (if they are internal). I was primarily targeting cleaning up the dead code from npymath in this PR. Maybe we should have a github project for code cleanup and add this as a task.

File: numpy/core/setup_common.py
Comment (mattip) [reply numpy/core/setup_common.py:146]: I think the shim part is already done, since cexp and cpow are not blocked for MSVC. Like glibc<2.18, many are blocked on MSVC because of accuracy problems. In any case, I will remove the comment.

File: numpy/core/src/common/npy_config.h
Comment (mattip) [reply numpy/core/src/common/npy_config.h:15]: Done.

**Comment by mattip** [comment]
I think this is a complete unit. There are at least two continuations:
- refactoring the complex math functions in a similar way: turning the macros from `HAVE_` to `BLOCK_` and making the never-blocked ones mandatory
- refactoring the mandatory functions so the `npy_wrapperfunc(*args) {return func(*args);};` are header-only

**Comment by rgommers** [comment]
> refactoring the complex math functions in a similar way: turning the macros from `HAVE_` to `BLOCK_` and making the never-blocked ones mandatory

May make sense, not entirely clear given that `complex.h` will remain optional forever it looks like. But either way, probably not in this PR?

**Comment by seberg** [comment]
I just browsed through and the changes look great.  I think I remember seeing that AIX might not define `nextafter` (not standard conform) due to using double-double.  But even if that is the case, I am +1 to going ahead with this and putting the definitions back if absolutely necessary.

File: numpy/core/src/npymath/npy_math_internal.h.src
Comment (rgommers) [inline numpy/core/src/npymath/npy_math_internal.h.src:324]: @mattip you removed `exp,log2,sqrt` from this list; they are handled later but without the `WORKAROUND_APPLE_TRIG_BUG` handling. I'm not sure if that was on purpose?

Comment (mattip) [reply numpy/core/src/npymath/npy_math_internal.h.src:324]: Yes it was on purpose. The comment about `WORKAROUND_APPLE_TRIG_BUG` says this, which only mentions `sin`, `cos`, `tan`. xref #19974 for more discussion.
> On arm64 macOS, there's a bug with sin, cos, and tan where they don't raise "invalid" when given INFINITY as input.

**Comment by mattip** [comment]
On a macmini running `% arch -arm64 python runtests.py -m slow` on main fails. Something about detecting the wrong long long type in dragon4.c: 
`numpy/core/src/multiarray/dragon4.c:3148:1: error: \
   implicit declaration of function 'Dragon4_PrintFloat_Intel_extended128' is \
   invalid in C99 [-Werror,-Wimplicit-function-declaration]`

**Comment by rgommers** [comment]
I'm not seeing that. Running `python runtests.py -m full` on macOS 12.4 (arm64) I only see these 3 (known, unrelated) failures:
```
FAILED numpy/f2py/tests/test_kind.py::TestKind::test_all - AssertionError: selectedrealkind(16): ex...
FAILED numpy/linalg/tests/test_linalg.py::TestDet::test_types[complex64] - RuntimeWarning: divide b...
FAILED numpy/linalg/tests/test_linalg.py::TestDet::test_types[complex128] - RuntimeWarning: divide ...
```

> `% arch -arm64 python`

I recommend not doing this, it is not necessary and running the non-native architecture like that has given me problems since the PowerPC OSX days.

**Comment by mattip** [comment]
In my m1 arm64 build using the system-provided compiler, `NPY_SIZEOF_LONGDOUBLE` is being reset here https://github.com/numpy/numpy/blob/acdd0e3ffff0db9f45951c6b51887937ca31935e/numpy/core/include/numpy/numpyconfig.h#L26-L28 which is wrong. `NPY_SIZEOF_LONGDOUBLE` should be 16. I wonder why you do not hit that in your build. Do you see the `__APPLE__` and `__arm64__` macros if you do
`touch /tmp/foo.c; clang -dM -E foo.c |grep "\(APPLE\)\|\(arm64\)"`?

**Comment by rgommers** [comment]
Yes:
```zsh
% touch /tmp/foo.c; clang -dM -E /tmp/foo.c |grep "\(APPLE\)\|\(arm64\)"
#define __APPLE_CC__ 6000
#define __APPLE__ 1
#define __arm64 1
#define __arm64__ 1
```

> which is wrong. `NPY_SIZEOF_LONGDOUBLE` should be 16.

I don't think so - `long double` is the same size as `double` on arm64, just like on `aarch64` and on Windows.

**Comment by rgommers** [comment]
This is a giveaway that you are not actually running the `arm64` binary that you think you are: `Dragon4_PrintFloat_Intel_extended128` (note _Intel_). If you cannot reproduce the error without using `-arch arm64 python`, then it's safe to ignore.

File: numpy/core/src/npymath/npy_math_internal.h.src
Comment (rgommers) [reply numpy/core/src/npymath/npy_math_internal.h.src:324]: Thanks, seems to work fine indeed.

**Comment by rgommers** [comment]
Tested on Linux as well, and verified that the number of defined symbols stayed unchanged (at 273) with: `nm -a libnpymath.a | grep "T npy_" | wc -l`

**Review by rgommers — APPROVED** [review]
Let's give this a go, thanks a lot @mattip!

**Comment by charris** [comment]
> I don't think so - long double is the same size as double on arm64, just like on aarch64

I believe the spec is 16 on aarch64, but Apple broke the spec for M1 and made it the same as double.

**Pull request merged**
